src/mongoc/mongoc-server-description.c

```diff
@@ -0,0 +1,483 @@
+ *
+ *-------------------------------------------------------------------------
+ */
+
+void
+mongoc_server_description_handle_ismaster (
+   mongoc_server_description_t   *sd,
+   const bson_t                  *ismaster_response,
+   int64_t                        rtt_msec,
+   bson_error_t                  *error)
+{
+   bson_iter_t iter;
+   bool is_master = false;
+   bool is_shard = false;
+   bool is_secondary = false;
+   bool is_arbiter = false;
+   /*
+   bool is_passive = false;
+   bool is_hidden = false;
+   */
+   bool is_replicaset = false;
+   const uint8_t *bytes;
+   uint32_t len;
+
+   bson_return_if_fail (sd);
+   bson_return_if_fail (ismaster_response);
+
+   bson_destroy (&sd->last_is_master);
+   bson_copy_to (ismaster_response, &sd->last_is_master);
+
+   bson_iter_init (&iter, &sd->last_is_master);
+
+   memset (&sd->set_name, 0, sizeof (*sd) - ((char*)&sd->set_name - (char*)sd));
+
+   while (bson_iter_next (&iter)) {
+      /* TODO: do we need to handle maxBsonObjSize */
+      if (strcmp ("ismaster", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_BOOL (&iter)) goto ERROR;
+         is_master = bson_iter_bool (&iter);
+      } else if (strcmp ("maxMessageSizeBytes", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_INT32 (&iter)) goto ERROR;
+         sd->max_write_batch_size = bson_iter_int32 (&iter);
+      } else if (strcmp ("minWireVersion", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_INT32 (&iter)) goto ERROR;
+         sd->min_wire_version = bson_iter_int32 (&iter);
+      } else if (strcmp ("maxWireVersion", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_INT32 (&iter)) goto ERROR;
+         sd->max_wire_version = bson_iter_int32 (&iter);
+      } else if (strcmp ("msg", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_BOOL (&iter)) goto ERROR;
+         is_shard = bson_iter_bool (&iter);
+      } else if (strcmp ("setName", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_UTF8 (&iter)) goto ERROR;
+
+         sd->set_name = bson_iter_utf8 (&iter, NULL);
+      } else if (strcmp ("secondary", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_BOOL (&iter)) goto ERROR;
+         is_secondary = bson_iter_bool (&iter);
+      } else if (strcmp ("hosts", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_ARRAY (&iter)) goto ERROR;
+         bson_iter_array (&iter, &len, &bytes);
+         bson_init_static (&sd->hosts, bytes, len);
+      } else if (strcmp ("passives", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_ARRAY (&iter)) goto ERROR;
+         bson_iter_array (&iter, &len, &bytes);
+         bson_init_static (&sd->passives, bytes, len);
+      } else if (strcmp ("arbiters", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_ARRAY (&iter)) goto ERROR;
+         bson_iter_array (&iter, &len, &bytes);
+         bson_init_static (&sd->arbiters, bytes, len);
```

Comment:
I'm going to hold off on this until SPEC-146 is resolved, I've put a note about it in the description above.